### PR TITLE
User can now create addresses.

### PR DESCRIPTION
**Pull request by samakshjain**
@sharoonthomas 
There are some minor issues, like country name is not displayed in the address list. I will fix it, but meanwhile please review the major changes.

### CODE REVIEW

File: shop/templates/layout.html [hunk 1]
```diff
@@ -17,10 +17,11 @@
 
   <!-- Mobile viewport optimized: h5bp.com/viewport -->
   <meta name="viewport" content="width=device-width">
-  
+
   <link href="https://fonts.googleapis.com/css?family=Montserrat" rel="stylesheet">
   <link href="https://fonts.googleapis.com/css?family=Loved+by+the+King" rel="stylesheet">
   <link rel="stylesheet" href="{{ url_for('static', filename='libs/font-awesome4/css/font-awesome.min.css') }}">
+  <script src="{{ url_for('static', filename='js/jquery.min.js') }}"></script>
```
Comment (sharoonthomas) [inline shop/templates/layout.html]: you should be using bower to get jquery. will update after merge

File: shop/templates/partials/my-account-menu.html [hunk 2]
```diff
@@ -10,14 +10,14 @@ <h4 class="currency-title">Select currency</h4>
 	      <li><a class="GBP"><span class="currencie-sym"><i class="fa fa-gbp"></i></span></a></li>
 	    </ul>
 	  </div>
-	        
+
 	  <div class="setting-option">
 	    <ul>
 	      <li><a href="">Wishlist</a></li>
-	      <li><a href="">MyAccount</a></li>
+	      <li><a href="">My Account</a></li>
```
Comment (sharoonthomas) [inline shop/templates/partials/my-account-menu.html]: 👍

File: shop/templates/users/addresses.html [hunk 3]
```diff
@@ -1,5 +1,32 @@
...
+    <h1>Address book</h1>
+    <br/>
+    <div class="container-fluid" style="margin-bottom: 30px">
+      <!-- Addresses -->
+      {% for address in addresses %}
+        <div class="row">
+          <div class="col-sm-6">
+            <address>
+              <strong>{{address.name}}</strong><br>
+                {{ address.street if address.street }}{{ address.streetbis if address.streetbis }}<br>
```
Comment (sharoonthomas) [inline shop/templates/users/addresses.html]: still not splitting full_address

File: shop/user/models.py [hunk 4]
```diff
@@ -139,3 +143,110 @@ def initiate_reset_password(self):
...
+    def get_addresses(cls):
+        """
+        Get all adresses of the user
+        """
+        return cls.query.filter_by_domain(
+            [('party', '=', current_user.party)]
+        ).all()
+
+    @classmethod
+    def get_address_form(cls, address=None):
```
Comment (sharoonthomas) [inline shop/user/models.py]: This is mixing model layer with view layer (forms). There should be clear separation of concerns.

File: shop/user/views.py [hunk 5]
```diff
@@ -15,4 +19,65 @@
 @login_required
 def addresses():
     """List Addresses."""
-    return render_template('users/addresses.html')
+    addresses = Address.get_addresses()
```
Comment (sharoonthomas) [inline shop/user/views.py]: I would prefer `current_user.get_addresses` (more readable and obvious)

File: shop/user/views.py [hunk 6]
```diff
@@ -15,4 +19,65 @@
...
+    ~~~
+    Return an address creation form
+    POST
+    ~~~~
+    Creates an address and redirects to the address view. If a next_url
+    is provided, redirects there.
+    """
+    form = Address.get_address_form()
+    if request.method == 'POST' and form.validate():
+        party = Party.get_by_id(current_user.party)
```
Comment (sharoonthomas) [inline shop/user/views.py]: why not `current_user.party.id`

File: shop/user/views.py [hunk 7]
```diff
@@ -15,4 +19,65 @@
...
+    GET
+    ~~~
+    Return an address creation form
+    POST
+    ~~~~
+    Creates an address and redirects to the address view. If a next_url
+    is provided, redirects there.
+    """
+    form = Address.get_address_form()
+    if request.method == 'POST' and form.validate():
```
Comment (sharoonthomas) [inline shop/user/views.py]: use `validate_on_submit()`

File: shop/user/views.py [hunk 8]
```diff
@@ -15,4 +19,65 @@
...
+        #     party.add_contact_mechanism_if_not_exists(
+        #         'phone', form.phone.data
+        #     )
+        return redirect(url_for('user.addresses'))
+    else:
+        flash_errors(form)
+
+    try:
+        return render_template('users/address-form.html', form=form)
+    except TemplateNotFound:
```
Comment (sharoonthomas) [inline shop/user/views.py]: what is this doing?

File: shop/user/views.py [hunk 9]
```diff
@@ -15,4 +19,65 @@
...
+        return render_template('users/address-form.html', form=form)
+
+
+@blueprint.route("/get_subdivision", methods=["GET"])
```
Comment (sharoonthomas) [inline shop/user/views.py]: take country as url argument

File: shop/user/views.py [hunk 10]
```diff
@@ -15,4 +19,65 @@
...
+@login_required
+def subdivision_list():
+    country = int(request.args.get('country', 0)) or 61
```
Comment (sharoonthomas) [inline shop/user/views.py]: not a good design pattern, but even if you were using it hardcoding an ID is bad.

```
country = request.args.get('country', type=int)
```

File: shop/user/views.py [hunk 11]
```diff
@@ -15,4 +19,65 @@
...
+    subdivisions = CountrySubdivision.get_filtered_subdivisions(country)
+    sub_dir = {"result": []}
```
Comment (sharoonthomas) [inline shop/user/views.py]: bad design. hard to read

File: shop/user/views.py [hunk 12]
```diff
@@ -15,4 +19,65 @@
...
+        return render_template('users/address-form.html', form=form)
+    except TemplateNotFound:
+        return render_template('users/address-form.html', form=form)
+
+
+@blueprint.route("/get_subdivision", methods=["GET"])
+@login_required
+def subdivision_list():
+    country = int(request.args.get('country', 0)) or 61
+    subdivisions = CountrySubdivision.get_filtered_subdivisions(country)
```
Comment (sharoonthomas) [inline shop/user/views.py]: I would rewrite the api to be:
```
country = Country(id=country_id)
subdivisions = country.subdivisions()
```

File: shop/templates/layout.html (hunk 1 above)
Comment (samakshjain) [inline shop/templates/layout.html]: Skipped it because I couldn't find a bower.json.

File: shop/templates/partials/my-account-menu.html (hunk 2 above)
Comment (samakshjain) [inline shop/templates/partials/my-account-menu.html]: OCD 😄

File: shop/templates/layout.html (hunk 1 above)
Comment (sharoonthomas) [inline shop/templates/layout.html]: in the root of the repo :)

File: shop/user/views.py (hunk 8 above)
Comment (samakshjain) [inline shop/user/views.py]: This was present in neried-webshop, and i figured checking and adding mechanism if DNE would be required here too. That't why I added a TODO tag.

File: shop/templates/layout.html (hunk 1 above)
Comment (samakshjain) [inline shop/templates/layout.html]: Woops

**Pull request closed without merging**